### PR TITLE
Feature Api Tasks App with Users 

### PR DESCRIPTION
**Pull request by AsierDev**
I've done the basics and added the possibility to assign tasks to users. Probably has many bugs, but if you don't stress it too much it should work :laughing:

### CODE REVIEW

File: staff/asier/tasks-list_express/src/data/tasksData.js [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+
+        if (task) return task
+
+        throw Error('Task does not exist.')
+    },
+
+    /*  /// Remove task from tasklist */
+
+    delete(id) {
+        const index = tasks.findIndex(x => x.taskId === Number(id))
```
Comment (Robertvera) [inline staff/asier/tasks-list_express/src/data/tasksData.js]: I would change param 'x' for 'task' for better syntax purpose

Comment (AsierDev) [reply staff/asier/tasks-list_express/src/data/tasksData.js]: You right! Changed... :+1:

File: staff/asier/tasks-list_express/src/data/tasksData.js [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+
+    deleteAll() {
+
+        tasks.splice(0, tasks.length)
+
+    },
+
+    /* /// Mark task as done  ///  */
+
+    markDone(id) {
```
Comment (manuelbarzi) [inline staff/asier/tasks-list_express/src/data/tasksData.js]: this is NOT a CONCERN of taskData, but of taskLogic. responsibility is mixed here. please, pay attention to the description of the challenge in trello ,)

File: staff/asier/tasks-list_express/src/data/tasksData.js [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+
+        if (tasks[index].done === true) throw Error('Task is already done')
+
+        tasks[index].done = true
+
+    },
+
+    /* /// List tasks pending to do /// */
+
+    listToDo() {
```
Comment (manuelbarzi) [inline staff/asier/tasks-list_express/src/data/tasksData.js]: this is NOT a CONCERN of taskData, but of taskLogic. responsibility is mixed here. please, pay attention to the description of the challenge in trello ,)

File: staff/asier/tasks-list_express/src/data/tasksData.js [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+        }
+
+        if (todoList < 1 || tasks.length < 1) throw Error('There are no tasks to show')
+
+        return todoList
+    },
+
+    /* /// List tasks already done /// */
+
+    listDone() {
```
Comment (manuelbarzi) [inline staff/asier/tasks-list_express/src/data/tasksData.js]: this is NOT a CONCERN of taskData, but of taskLogic. responsibility is mixed here. please, pay attention to the description of the challenge in trello ,)

File: staff/asier/tasks-list_express/src/data/tasksData.js [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+
+            task.text = text
+        }
+
+    },
+
+    /* /// List tasks per user /// */
+
+    userTasks(username) {
+        const userList = []
```
Comment (manuelbarzi) [inline staff/asier/tasks-list_express/src/data/tasksData.js]: this is NOT a CONCERN of taskData, but of an hypothetical userData that is not part of this challenge. responsibility is mixed here. please, pay attention to the description of the challenge in trello ,)

File: staff/asier/tasks-list_express/src/data/tasksData.js [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+
+        if (userList < 1 || tasks.length < 1) throw Error('There are no tasks to show')
+
+        return userList
+    },
+
+    /*  /// List all the users /// */
+
+    listUsers() {
+        if (users.length < 1) throw Error('There are no users to show')
```
Comment (manuelbarzi) [inline staff/asier/tasks-list_express/src/data/tasksData.js]: this is NOT a CONCERN of taskData, but of an hypothetical userData that is not part of this challenge. responsibility is mixed here. please, pay attention to the description of the challenge in trello ,)

File: staff/asier/tasks-list_express/src/logic/tasksLogic.js [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+
+    remove(id) {
+
+        tasksData.delete(id)         
+    },
+
+    /* /// List tasks already done /// */
+
+    listDone() {
+        return tasksData.listDone()
```
Comment (manuelbarzi) [inline staff/asier/tasks-list_express/src/logic/tasksLogic.js]: taskLogic MUST do the logic contained inside this method from taskData, but not the taskData itself. responsibility is mixed here. please, pay attention to the description of the challenge in trello ,)

File: staff/asier/tasks-list_express/src/logic/tasksLogic.js [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+    },
+
+    /* /// List tasks pending to do /// */
+
+    listToDo() {
+        return tasksData.listToDo()
```
Comment (manuelbarzi) [inline staff/asier/tasks-list_express/src/logic/tasksLogic.js]: taskLogic MUST do the logic contained inside this method from taskData, but not the taskData itself. responsibility is mixed here. please, pay attention to the description of the challenge in trello ,)

File: staff/asier/tasks-list_express/src/logic/tasksLogic.js [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+            validateText(text)
+            validateUsername(username)
+        }
+        
+    },
+
+    /* /// List tasks per user /// */
+
+    userTasks(username) {
+        return tasksData.userTasks(username)
```
Comment (manuelbarzi) [inline staff/asier/tasks-list_express/src/logic/tasksLogic.js]: this is NOT part of the challenge. the idea was just to add a field called username in a task, but not to manage users. managing users is not part of task api.

**Comment by AsierDev** [comment]
Since I got the concept of users wrong, I'm closing this pr to make everything again from new

**Comment by manuelbarzi** [comment]
NO need to CLOSE the PR, @AsierDev, just attending the comments and applying the changes to update the code is enough. PRs are exactly for that, not for just making a try and then closing them if something is not accurate. the idea is to learn from and in them, and then apply the changes as necessary, attending the reviews in comments. PRs are organic, they are subject to code reviews and "updation" ) then iterating over them, improving the code, but not closing them as if they are not valid because of needing some little adjustments. 

regarding, layers and resposibilities, check this PR github.com/manuelbarzi/skylab-bootcamp-201801/pull/79 the code seems to be quite accurate to what expected. then you would just need to do a few adjustments to get your code done ,)

**Comment by AsierDev** [comment]
Ah ok then, I'll do that. I'll push it later

**Comment by AsierDev** [comment]
Hopefully now is better. I've adapted it (I think) to what you asked. I didn't include the users this time.

File: staff/asier/tasks-list_express/src/data/tasksData.js (hunk 6 above)
Comment (AsierDev) [reply staff/asier/tasks-list_express/src/data/tasksData.js]: done

File: staff/asier/tasks-list_express/src/data/tasksData.js (hunk 5 above)
Comment (AsierDev) [reply staff/asier/tasks-list_express/src/data/tasksData.js]: Got rid of it

File: staff/asier/tasks-list_express/src/data/tasksData.js [hunk 10]
```diff
@@ -0,0 +1,86 @@
+const tasks = []
+
+let taskId = 0
+
+const tasksData = {
+
+    /* /// create task /// */
+
+    create(text) {
```
Comment (manuelbarzi) [inline staff/asier/tasks-list_express/src/data/tasksData.js]: ADD done param

File: staff/asier/tasks-list_express/src/logic/tasksLogic.js [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+
+    removeAll() {
+        tasksData.deleteAll()
+    },
+
+    /* /// Update task /// */
+
+    update(id, text, done) {
+
+        tasksData.update(id, text)
```
Comment (manuelbarzi) [inline staff/asier/tasks-list_express/src/logic/tasksLogic.js]: TRANSFER done to taskData

**Comment by AsierDev** [comment]
Changes made :+1:

**Comment by manuelbarzi** [comment]
WELL done ,)

**Pull request merged**